ui: multi-select and batched gallery image operations

psychedelicious opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission


## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No 😬 


## Description

[feat: add multi-select to gallery](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)
[c3e99ad](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)

multi-select actions include:
- drag to board to move all to that board
- right click to add all to board or delete all

backend changes:
- add routes for changing board for list of image names, deleting list of images
- change image-specific routes to `images/i/{image_name}` to not clobber other routes (like `images/upload`, `images/delete`)
- subclass pydantic `BaseModel` as `BaseModelExcludeNull`, which excludes null values when calling `dict()` on the model. this fixes inconsistent types related to JSON parsing null values into `null` instead of `undefined`
- remove `board_id` from `remove_image_from_board`

frontend changes:
- multi-selection stuff uses `ImageDTO[]` as payloads, for dnd and other mutations. this gives us access to image `board_id`s when hitting routes, and enables efficient cache updates.
- consolidate change board and delete image modals to handle single and multiples
- board totals are now re-fetched on mutation and not kept in sync manually - was way too tedious to do this
- fixed warning about nested `<p>` elements
- closes #4088 , need to handle case when `autoAddBoardId` is `"none"`
- add option to show gallery image delete button on every gallery image

frontend refactors/organisation:
- make typegen script js instead of ts
- enable `noUncheckedIndexedAccess` to help avoid bugs when indexing into arrays, many small changes needed to satisfy TS after this
- move all image-related endpoints into `endpoints/images.ts`, its a big file now, but this fixes a number of circular dependency issues that were otherwise felt impossible to resolve

## Related Tickets & Documents

- Supersedes #3669 #3771
- Closes #3882 #4088

## QA Instructions, Screenshots, Recordings

Basic functionality to review:
- Use shift and ctrl/cmd to select multiple images
- Drag and drop images between boards
- Right click multi-selection to change the board for a group of images
- Right click multi-selection to delete many images at once
- General drive-through of gallery features & auto-add

Specific cases:
- Deleting a single image should still work as before
- Deleting a single image that is in use in another feature should prompt you and reset that feature when deletion is confirmed
- Same as above for multiple images
- Adding a single image to a board should still work as before
- Adding multiple images to a board should work via context menu and dnd

There's a lot to test here and some complicated cache updates, so it's hard to test all the possible sequences of events.

<!-- 
Please provide steps on how to test changes, any hardware or 
software specifications as well as any other pertinent information. 
-->

## Added/updated tests?

- [ ] Yes
- [x] No 😬 

## [optional] Are there any post deployment tasks we need to perform?

psychedelicious commented on invokeai/app/api/routers/board_images.py:
@maryhipp These are the changes needed for batch `board_image` operations. May be some additional changes but just giving you a heads up. The destination `board_id` is provided in each request - do we also need to provide the image's source/current `board_id`?

psychedelicious replied:
Ok, would it be helpful to provide a original board ids with their image names though? 

Good that it's not necessary but we now have access to them when request is made, so it's no trouble to include them.

psychedelicious replied:
Sorry to back and forth on this. In the end, I've removed `board_id` from the `remove from board` routes and services.

psychedelicious commented on invokeai/frontend/web/src/services/api/endpoints/boards.ts:
```diff
@@ -82,6 +83,44 @@ export const boardsApi = api.injectEndpoints({
     }),
 
+    getBoardImagesTotal: build.query<number, string | undefined>({
+      query: (board_id) => ({
+        url: getListImagesUrl({
+          board_id: board_id ?? 'none',
+          categories: IMAGE_CATEGORIES,
+          is_intermediate: false,
+          limit: 0,
+          offset: 0,
+        }),
+        method: 'GET',
+      }),
+      providesTags: (result, error, arg) => [
+        { type: 'BoardImagesTotal', id: arg ?? 'none' },
+      ],
+      transformResponse: (response: OffsetPaginatedResults_ImageDTO_) => {
+        return response.total;
+      },
+    }),
+
+    getBoardAssetsTotal: build.query<number, string | undefined>({
+      query: (board_id) => ({
+        url: getListImagesUrl({
+          board_id: board_id ?? 'none',
+          categories: ASSETS_CATEGORIES,
+          is_intermediate: false,
+          limit: 0,
+          offset: 0,
+        }),
+        method: 'GET',
+      }),
+      providesTags: (result, error, arg) => [
+        { type: 'BoardAssetsTotal', id: arg ?? 'none' },
+      ],
+      transformResponse: (response: OffsetPaginatedResults_ImageDTO_) => {
+        return response.total;
+      },
+    }),
+
```
I tried manually keeping the board totals in sync but boy was it a pain. I'm just using the list images routes, with a `limit` of `0`, as a easy way to query for a board's image counts. 

We need separate queries for the `images` and `assets`, so while we could make dedicated routes for the counts, we'd still have to do a join. So I figure using the existing route is fine for now.

blessedcoolant replied:
This man reviewing his own PR. We can call it a day. :P

psychedelicious replied:
Lol

psychedelicious commented on invokeai/frontend/web/tsconfig.json:
```diff
@@ -13,6 +13,8 @@
     "moduleResolution": "Node",
     // TODO: Disabled for IDE performance issues with our translation JSON
     // "resolveJsonModule": true,
+    "noUncheckedIndexedAccess": true,
+    "strictNullChecks": true,
```
Tightened up the reins a bit. Ran into some bugs related to indexed array access: TS does not type indexed array access values as a union with `undefined`, so it's very easy to end up with undefined when working with arrays.

Enabling this rule unions *all* indexed array access values with `undefined`. Many more runtime checks are needed, but no more errors.

psychedelicious commented on invokeai/frontend/web/src/app/components/ImageDnd/typesafeDnd.tsx:
```diff
@@ -78,14 +82,12 @@ export type ImageDraggableData = BaseDragData & {
   payload: { imageDTO: ImageDTO };
 };
 
-export type ImageNamesDraggableData = BaseDragData & {
-  payloadType: 'IMAGE_NAMES';
-  payload: { image_names: string[] };
+export type ImageDTOsDraggableData = BaseDragData & {
+  payloadType: 'IMAGE_DTOS';
+  payload: { imageDTOs: ImageDTO[] };
```
I changed the drag and drop payload for multiple images to be an array of DTOs instead of just image names. This gives us more flexibility when handling these dnd events. 

For example, we now have access to an image's current `board_id` when dropping multiple images on a board. This allows us to greatly reduce the number of network requests needed when doing batched board management operations.

psychedelicious commented on invokeai/frontend/web/scripts/typegen.js:
There's no need for this to be a TS script. Changed to JS, runs a lot faster now (not like it was slow before tho).

psychedelicious commented on invokeai/app/util/model_exclude_null.py:
Weird issue with the OpenAPI types - the `ImageDTO`s were coming back with `null` as values for empty properties. For example, if an image was not on a board, it would have `board_id: null` instead of `board_id: undefined` or not having a `board_id` at all.

The generated types for eg `ImageDTO` do *not* have `null` in them anywhere - so we were expecting `board_id: string | undefined` but the actual object were `board_id: string | null`. This caused *really* confusing problems.

The solution is to override the pydantic `BaseModel`'s `dict()` method to use the `exclude_none` argument. Then the `ImageDTO`s simply omit empty props, and the objects we receive have the expected shape.

Instead of overriding `dict()` on every model that needs it, we can subclass `BaseModel` and have eg `ImageDTO` inherit from it.

maryhipp commented on invokeai/app/api/routers/board_images.py:
```diff
@@ -53,3 +67,66 @@ async def remove_board_image(
+)
+async def create_board_images_from_list(
+    board_images: list[BoardImage] = Body(
+        description="The board images to create", embed=True
```
What is the use case for having a list of BoardImages vs a list of Images with one board_id? Are there front-end actions that can trigger this for multiple board/image combinations?

psychedelicious replied:
Yes, a batch may have images from any board, and this allows it to support multi select in an intuitive way. Imagine you want to make a board out of your current batch

psychedelicious replied:
I decided to just make this a list of image names with a single `board_id`, we can handle the other behavior if the actual need arises

maryhipp commented:
I have only skimmed this so far, but wondering what happens when a group of images are moved from board A to board B? I only see functionality to add or remove from board, not change boards.

psychedelicious commented:
@maryhipp Because images can only be on one board, creating a board image is equivalent to both moving an image from one board to another, and adding image to a board. 

We could add move as a distinct operation but there's substantially more logic to add, and it makes multiple image operations a lot more complex.

maryhipp commented:
> @maryhipp Because images can only be on one board, creating a board image is equivalent to both moving an image from one board to another, and adding image to a board.
> 
> We could add move as a distinct operation but there's substantially more logic to add, and it makes multiple image operations a lot more complex.

OK - I am a bit wary of relying solely on a database constraint to dictate what is essentially business logic (that is my understanding of the implementation); I feel like it could lead to unexpected or unintended behavior if someone isn't aware of it or doesn't understand SQL. This is the only line that is enforcing this, right? `ON CONFLICT (image_name) DO UPDATE SET board_id = ?;`

psychedelicious commented:
> OK - I am a bit wary of relying solely on a database constraint to dictate what is essentially business logic (that is my understanding of the implementation); I feel like it could lead to unexpected or unintended behavior if someone isn't aware of it or doesn't understand SQL. This is the only line that is enforcing this, right? `ON CONFLICT (image_name) DO UPDATE SET board_id = ?;`

Yep that's what handles changing the board if the image is already on another board. A more explicit handling might be to first remove images from their board before adding them, in the `add_images_to_board` route. Should behave exactly the same

psychedelicious commented:
Ok, ready for review. Squashed:

[feat: add multi-select to gallery](https://github.com/invoke-ai/InvokeAI/pull/4032/commits/c3e99ad145e73a5d836300c70bd81aeb5b574a7c)

multi-select actions include:
- drag to board to move all to that board
- right click to add all to board or delete all

backend changes:
- add routes for changing board for list of image names, deleting list of images
- change image-specific routes to `images/i/{image_name}` to not clobber other routes (like `images/upload`, `images/delete`)
- subclass pydantic `BaseModel` as `BaseModelExcludeNull`, which excludes null values when calling `dict()` on the model. this fixes inconsistent types related to JSON parsing null values into `null` instead of `undefined`
- remove `board_id` from `remove_image_from_board`

frontend changes:
- multi-selection stuff uses `ImageDTO[]` as payloads, for dnd and other mutations. this gives us access to image `board_id`s when hitting routes, and enables efficient cache updates.
- consolidate change board and delete image modals to handle single and multiples
- board totals are now re-fetched on mutation and not kept in sync manually - was way too tedious to do this
- fixed warning about nested `<p>` elements
- closes https://github.com/invoke-ai/InvokeAI/issues/4088 , need to handle case when `autoAddBoardId` is `"none"`
- add option to show gallery image delete button on every gallery image

frontend refactors/organisation:
- make typegen script js instead of ts
- enable `noUncheckedIndexedAccess` to help avoid bugs when indexing into arrays, many small changes needed to satisfy TS after this
- move all image-related endpoints into `endpoints/images.ts`, its a big file now, but this fixes a number of circular dependency issues that were otherwise felt impossible to resolve

psychedelicious commented:
I've updated the PR description with changes and how to test.

maryhipp commented on invokeai/app/api/routers/images.py:
```diff
@@ -77,7 +78,7 @@ async def upload_image(
 
 
-@images_router.delete("/{image_name}", operation_id="delete_image")
+@images_router.delete("/i/{image_name}", operation_id="delete_image")
```
Why these routes changes?

psychedelicious replied:
Concerned that an image name may clobber `images/upload` or `images/batch` (or any other routes we may add in the future). It really shouldn't happen bc the image names are UUIDv4, but this makes it impossible

maryhipp commented on invokeai/frontend/web/src/features/controlNet/components/parameters/ParamControlNetWeight.tsx:
```diff
@@ -1,32 +1,18 @@
 
 const ParamControlNetWeight = (props: ParamControlNetWeightProps) => {
-  const { controlNetId } = props;
+  const { weight, isEnabled, controlNetId } = props.controlNet;
```
All of these controlnet config changes are unrelated to multiselect, right?

psychedelicious replied:
Yes - I kept on running into bugs related to indexed array access that gave me `undefined` when I expected an `ImageDTO`, due to the array length. Typescript does not mark indexed access as possibly undefined by default.

I enabled `noUncheckedIndexedAccess` in `tsconfig` to make Typescript be stricter with this, which showed me all the places where this class of but could pop up. 

Unfortunately, it was a lot of places unrelated to the gallery, and I fixed all occurrences.

maryhipp commented:
I just tested locally and it looks great! Everything seems to be working really well. The only thing I noticed is that when you choose multiple images to delete using multiselect, there is a lengthy pause (this was with like 20 images) in the UI before anything happens. Maybe a small loading state or something would help, but just a nice-to-have.

maryhipp reviewed (approved):
Had some small questions but overall this looks 🔥 and I think we can always iterate on the small stuff

psychedelicious commented:
> I just tested locally and it looks great! Everything seems to be working really well. The only thing I noticed is that when you choose multiple images to delete using multiselect, there is a lengthy pause (this was with like 20 images) in the UI before anything happens. Maybe a small loading state or something would help, but just a nice-to-have.

Assuming you mean after you confirm a multiple Delete - I couldn't figure out how to do optimistic updates for these actions without totally invalidating the caches, so we are waiting for the server response before updating the UI. That's the delay

This is the only un-ticked checkbox on the roadmap for this feature... we'd need to "freeze" all images pending mutation (hard) or freeze all image interactions (ugly) while we wait for the network request to finish. If a pause is tenable for now, I'd like to address this in a followup.

maryhipp commented:
We could also just add a loader or something over the gallery during that delay - but agree it can be a follow-up

brandonrising commented on invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/imageDeleted.ts:
```diff
@@ -1,57 +1,72 @@
 
-      dispatch(isModalOpenChanged(false));
+      if (imageDTOs.length !== 1 || imagesUsage.length !== 1) {
+        // handle multiples in separate listener
```
Any reason single and multi image deletion couldn't be the same listener and backend endpoint? Like, single image could just send a list of images of size 1. Just seems like this could be annoying to maintain(or forgotten about) in the future.

psychedelicious replied:
I had approached the same conclusion but am a bit burnt out on the gallery stuff and would prefer to leave it for a future cleanup

psychedelicious replied:
Raised in  #4142

brandonrising commented on invokeai/frontend/web/src/features/canvas/hooks/useColorUnderCursor.ts:
```diff
@@ -33,6 +33,10 @@ const useColorPicker = () => {
           1
         ).data;
 
+      if (!(a && r && g && b)) {
```
Even when it makes sense for variables to be 1 letter long, it still so PTSD inducing from my university math days lol

psychedelicious replied:
I promise there aren't many of these! This is only in the PR due to typescript refactoring, else you'd never have known

brandonrising reviewed (approved):
Overall lgtm, I tried pretty much everything in the UI that's affected by these changes and it all looked great. I left a few comments but I don't think any of them are blocking, feel free to ignore them or address them in a future PR. Great work!

The only other high level thing I thought of while reviewing this is that multi-select download could be really nice to have as well.

psychedelicious commented on invokeai/app/api/routers/board_images.py:
```diff
@@ -29,26 +35,78 @@ async def create_board_image(
+                )
+                added_image_names.append(image_name)
+            except:
+                pass
```
Yeah, probably so. Raised in #4141

psychedelicious commented:
I've resolved merge conflicts related to #4086 and tested that the auto-assign on click still works. Thanks for the thorough review @brandonrising !
